src/client/pythonEnvironments/collection/environmentsReducer.ts

```diff
@@ -0,0 +1,110 @@
+import { mergeEnvironments } from '../info';
+
+type PythonEnvsIteratorState = {
+    /**
+     * Set to true when all information from incoming iterator has been received
+     */
+    done: boolean;
+    /**
+     * Carries the number of pending background calls ongoing at the moment
+     */
+    pending: number;
+};
+
+export class PythonEnvsReducer implements ILocator {
+    public get onChanged(): Event<PythonEnvsChangedEvent> {
+        return this.pythonEnvsManager.onChanged;
+    }
+
+    constructor(private readonly pythonEnvsManager: ILocator) {}
+
+    public resolveEnv(env: string | PythonEnvInfo): Promise<PythonEnvInfo | undefined> {
+        return this.pythonEnvsManager.resolveEnv(env);
+    }
+
+    public iterEnvs(query?: QueryForEvent<PythonEnvsChangedEvent>): IPythonEnvsIterator {
+        const didUpdate = new EventEmitter<PythonEnvUpdatedEvent | null>();
+        return { ...this.iterEnvsIterator(didUpdate, query), onUpdated: didUpdate.event };
+    }
+
+    private async* iterEnvsIterator(
+        didUpdate: EventEmitter<PythonEnvUpdatedEvent | null>,
+        query?: QueryForEvent<PythonEnvsChangedEvent>,
+    ): AsyncIterator<PythonEnvInfo, void> {
+        const state: PythonEnvsIteratorState = {
+            done: false,
+            pending: 0,
+        };
+        const seen: PythonEnvInfo[] = [];
+        const iterator = this.pythonEnvsManager.iterEnvs(query);
+
+        if (iterator.onUpdated !== undefined) {
+            iterator.onUpdated((event) => {
+                if (event === null) {
+                    state.done = true;
+                    checkIfFinishedAndNotify(state, didUpdate);
+                } else {
+                    const old = seen.find((s) => s.location === event.old.location);
+                    if (old !== undefined) {
+                        state.pending += 1;
+                        resolveDifferencesInBackground(old, event.new, state, didUpdate).ignoreErrors();
+                    }
+                }
+            });
+        }
+
+        let result = await iterator.next();
+        while (!result.done) {
+            const currEnv = result.value;
+            const old = seen.find((s) => s.location === currEnv.location);
+            if (old !== undefined) {
+                state.pending += 1;
+                resolveDifferencesInBackground(old, currEnv, state, didUpdate).ignoreErrors();
+            } else {
+                yield currEnv;
+                seen.push(currEnv);
+            }
+            // eslint-disable-next-line no-await-in-loop
+            result = await iterator.next();
+        }
+        if (iterator.onUpdated === undefined) {
+            state.done = true;
+        }
+    }
+}
+
+async function resolveDifferencesInBackground(
+    oldEnv: PythonEnvInfo,
+    newEnv: PythonEnvInfo,
+    state: PythonEnvsIteratorState,
+    didUpdate: EventEmitter<PythonEnvUpdatedEvent | null>,
+) {
+    const merged = mergeEnvironments([oldEnv, newEnv])[0];

```

Comment:
Please clarify why we need `mergeEnvironments()` to so complex.  I would expect nothing more than:

```suggestion
    const merged = mergeEnvironments(oldEnv, newEnv);
```

Reply:
The job of the "reducer" class is to resolve collisions and the most important part of that is to determine which env kind to use when they are different.  So we need logic to make sure we select the proper kind.  At the least I'd expect this:

```suggestion
    const merged = mergeEnvironments([oldEnv, newEnv])[0];
    merged.kind = decideEnvKind(oldEnv, newEnv);
```

and we may need to adjust other merged data based on the decided kind:

```suggestion
    const kind = decideEnvKind(oldEnv, newEnv);
    const merged = mergeEnvsAsKind(kind, oldEnv, newEnv);
```

or even:

```suggestion
    const merged = resolveEnvCollision(oldEnv, newEnv);
```

Reply:
Perhaps there is an expectation that the env kind will always be the same when there is a collision?  There is no guarantee of that though.

Reply:
> Please clarify why we need mergeEnvironments() to so complex

I had clarified it before here https://github.com/microsoft/vscode-python/pull/13953#discussion_r490173100 Looks like you missed it. Essential this is a general merge method [which already exists](https://github.com/Microsoft/vscode-python/blob/e7d7fdc13c11044fc7de4940e5e40b922aa33da7/src/client/pythonEnvironments/info/index.ts#L183) , Karthik is working on changing to support the new type.

> Perhaps there is an expectation that the env kind will always be the same when there is a collision? There is no guarantee of that though.

That merge functions will take care of which kind to choose in case there're multiple types.